ledger/participant-state/kvutils/tools/integrity-check/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/WriteItemParser.scala

```diff
@@ -0,0 +1,37 @@
+package com.daml.ledger.participant.state.kvutils.tools.integritycheck
+
+import com.daml.ledger.participant.state.kvutils.`export`.WriteItem
+import com.daml.ledger.participant.state.kvutils.{DamlKvutils, Envelope, Raw}
+import com.daml.ledger.validator.StateKeySerializationStrategy
+
+class WriteItemParser(stateKeySerializationStrategy: StateKeySerializationStrategy) {
+
+  def parse(writeItem: WriteItem): Either[String, Either[
+    (DamlKvutils.DamlLogEntryId, DamlKvutils.DamlLogEntry),
+    (DamlKvutils.DamlStateKey, DamlKvutils.DamlStateValue),
+  ]] = {
+    val (rawKey, rawEnvelope) = writeItem
+    Envelope.open(rawEnvelope) match {
+      case Left(errorMessage) =>
+        Left(s"Invalid value envelope: $errorMessage")
+      case Right(Envelope.LogEntryMessage(logEntry)) =>
+        val logEntryId = DamlKvutils.DamlLogEntryId.parseFrom(rawKey.bytes)
+        if (logEntry.getPayloadCase == DamlKvutils.DamlLogEntry.PayloadCase.PAYLOAD_NOT_SET)
+          Left("Log entry payload not set.")
+        else
+          Right(Left(logEntryId -> logEntry))
+      case Right(Envelope.StateValueMessage(value)) =>
+        val key = stateKeySerializationStrategy.deserializeStateKey(Raw.StateKey(rawKey.bytes))
+        if (key.getKeyCase == DamlKvutils.DamlStateKey.KeyCase.KEY_NOT_SET)
+          Left("State key not set.")
+        else if (value.getValueCase == DamlKvutils.DamlStateValue.ValueCase.VALUE_NOT_SET)
+          Left("State value not set.")
+        else
+          Right(Right(key -> value))
+      case Right(Envelope.SubmissionMessage(submission)) =>
+        Left(s"Unexpected submission message: $submission")
```

Comment:
I'm not yet sure why Submission messages are unexpected here and how to handle all of the types. I think you @miklos-da have been referring to Submission's fields.